Source/DafnyCore/AST/Members/Function.cs

```diff
@@ -74,6 +74,10 @@ public class Function : MemberDecl, TypeParameter.ParentType, ICallable, ICanFor
       yield return new Assumption(this, tok, AssumptionDescription.ExternWithPrecondition);
     }
 
+    if (Attributes.Contains(Reads.Attributes, Attributes.AssumeConcurrentAttributeName)) {
```

Comment:
Because functions always rely on all their reads to be immutable from verification point of view, I would actually remove the ability to put {:concurrent} in front of a function.
Methods, on the other hand, can encode non determinism by calling other methods, so for concurrency support, I'd remove functions as potential entry point altogether to avoid pitfalls.

Reply:
I'm totally willing to only support `{:assume_concurrent}` on methods, since as you say the risk of allowing non-empty `reads` and `modifies` on functions is greater, and I don't know of any use cases that want it on functions (e.g. `smithy-dafny` never generates any function entry points with `reads` clauses).

But I don't see any reason to not allow a `{:concurrent} function` with an empty `reads` clause, and `smithy-dafny` does generate `function` entry points for operations with the `@readonly` trait.